src/datanode/src/heartbeat.rs

```diff
@@ -101,39 +123,98 @@ impl HeartbeatTask {
         let node_id = self.node_id;
         let addr = resolve_addr(&self.server_addr, &self.server_hostname);
         let meta_client = self.meta_client.clone();
-
         let catalog_manager_clone = self.catalog_manager.clone();
-        let mut tx = Self::create_streams(&meta_client, running.clone()).await?;
+        let handler_executor = self.heartbeat_response_handler_exector.clone();
+
+        let (outgoing_tx, mut outgoing_rx) = mpsc::channel(16);
+        let mailbox = Arc::new(HeartbeatMailbox::new(outgoing_tx));
+
+        let mut tx = Self::create_streams(
+            &meta_client,
+            running.clone(),
+            handler_executor.clone(),
+            mailbox.clone(),
+        )
+        .await?;
+
         common_runtime::spawn_bg(async move {
-            while running.load(Ordering::Acquire) {
-                let (region_num, region_stats) = datanode_stat(&catalog_manager_clone).await;
-
-                let req = HeartbeatRequest {
-                    peer: Some(Peer {
-                        id: node_id,
-                        addr: addr.clone(),
-                    }),
-                    node_stat: Some(NodeStat {
-                        region_num: region_num as _,
-                        ..Default::default()
-                    }),
-                    region_stats,
-                    ..Default::default()
-                };
+            let sleep = tokio::time::sleep(Duration::from_millis(0));
+            tokio::pin!(sleep);
 
-                if let Err(e) = tx.send(req).await {
-                    error!("Failed to send heartbeat to metasrv, error: {:?}", e);
-                    match Self::create_streams(&meta_client, running.clone()).await {
-                        Ok(new_tx) => {
-                            info!("Reconnected to metasrv");
-                            tx = new_tx;
+            loop {
+                if !running.load(Ordering::Acquire) {
+                    info!("shutdown heartbeat task");
+                    break;
+                }
+                let req = tokio::select! {
+                    message = outgoing_rx.recv() => {
+                        if let Some(message) = message {
+                            match outgoing_message_to_mailbox_message(message) {
+                                Ok(message) => {
+                                    let (region_num, region_stats) = datanode_stat(&catalog_manager_clone).await;
```

Comment:
Do we still need to collect datanode stat when sending mailbox msg? I prefer not, to achieve more clear codes. cc @fengjiachun

Reply:
+1